orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/util/PackageInfo.groovy

```diff
@@ -215,4 +228,20 @@ class PackageInfo {
     }
   }
 
+  @CompileDynamic
+  private Map findBuildInfoInUpstreamStage() {
+
+    stage?.ancestors()?.findAll {
```

Comment:
Maybe something like ...

```
def upstreamStage = stage.ancestors().find { 
  artifactMatch(it.stage.context.buildInfo?.artifacts as List<Map<String, String>)
}?.stage
return upstreamStage ? upstreamStage.context.buildInfo as Map : null
```

If you made this method static and passed in the `Stage`, you could unit it test it independently of anything else.